### PR TITLE
Reintroduce standing matrix

### PR DESCRIPTION
**Pull request by dmarcos**
The Vive examples are currently broken. The assigned user height from the camera rig adds up to the height coming from the VRDisplay pose resulting in a camera that it's much higher than expected. Oculus and Vive tracking systems are different. Oculus Rift does not have the notion of ground and the position is always relative to where the sensors start reading or are reset by the user (The same applies to Windows MR). On the other hand, Vive knows the position relative to the ground and you always get the user real height. A way to normalize this is always consider standing coordinates by applying the sittingToStandingTransform. I never fully understood the purpose of sitting experiences. In standing coordinates you can use Oculus both sitting and standing. I don't see any limitations when it comes to content creation. By always transforming to standing coordinates both Oculus and Vive behave the same way for the content developer. The only difference is that Vive returns the real height of the user thanks to Lighthouse tracking capabilities and Oculus calculates the height based on the value introduced by the user when calibrating the headset. Tested on:

- [x] Windows Mixed Reality
- [x] Oculus Rift
- [x] Vive

### CODE REVIEW

**Comment by dmarcos** [comment]
This will fix https://github.com/mrdoob/three.js/issues/11619

**Review by Mugen87 — APPROVED** [review]

**Comment by arturitu** [comment]
I was testing with Microsoft Mixed Reality portal / Edge and headset is at y = 0 (It doesn't receive standingMatrix, only pose.position based on the initial headset position that is 0,0,0)
capturedraggingmrp
I think we will need to add an else on 'if ( stageParameters ) {' to add the scope.userHeight if this info is not coming from the system.

File: src/renderers/webvr/WebVRManager.js [hunk 1]
```diff
@@ -121,6 +126,18 @@ function WebVRManager( renderer ) {
...
+		if ( stageParameters ) {
+
+			standingMatrix.fromArray( stageParameters.sittingToStandingTransform );
+			standingMatrixInverse.getInverse( standingMatrix );
+		
+			poseObject.matrixWorld.multiply( standingMatrix );
+			camera.matrixWorldInverse.multiply( standingMatrixInverse );
+		
+		}
+
```
Comment (fernandojsg) [inline src/renderers/webvr/WebVRManager.js]: If no `stageParameters` we should create a `standingMatrix` just with `translate(0, scope.userHeight, 0)` and apply to the poseObject and camera too.

**Review by fernandojsg — CHANGES_REQUESTED** [review]

File: src/renderers/webvr/WebVRManager.js (hunk 1 above)
Comment (fernandojsg) [reply src/renderers/webvr/WebVRManager.js]: Something like:
```
if (stageParameters) {
  standingMatrix.fromArray( stageParameters.sittingToStandingTransform );
  standingMatrixInverse.getInverse( standingMatrix );
        
  poseObject.matrixWorld.multiply( standingMatrix );
  camera.matrixWorldInverse.multiply( standingMatrixInverse );
} else {
  standingMatrix.identity().makeTranslation(0, scope.userHeight, 0);
  standingMatrixInverse.getInverse( standingMatrix );
        
  poseObject.matrixWorld.multiply( standingMatrix );
  camera.matrixWorldInverse.multiply( standingMatrixInverse );
}
```
I didn't test it though

File: src/renderers/webvr/WebVRManager.js [hunk 2]
```diff
@@ -137,6 +154,11 @@ function WebVRManager( renderer ) {
 		cameraL.matrixWorldInverse.fromArray( frameData.leftViewMatrix );
 		cameraR.matrixWorldInverse.fromArray( frameData.rightViewMatrix );
 
+		if ( stageParameters ) {
```
Comment (fernandojsg) [inline src/renderers/webvr/WebVRManager.js]: As we already will have an `standingMatrix` no matter if from `stageParameters` or created using `userHeight` this should not be an if()

**Review by fernandojsg — CHANGES_REQUESTED** [review]
We need a `standingmatrix` applied with  `userHeight` even if there is no `stageParameters`.

**Comment by arturitu** [comment]
Now with the @fernandojsg requested changes is working ok (at userHeight from the floor) with MR headset too.

File: src/renderers/webvr/WebVRManager.js (hunk 1 above)
Comment (fernandojsg) [reply src/renderers/webvr/WebVRManager.js]: Well... you know
```
if (stageParameters) {
  standingMatrix.fromArray( stageParameters.sittingToStandingTransform );
} else {
  standingMatrix.identity().makeTranslation(0, scope.userHeight, 0);
}
  standingMatrixInverse.getInverse( standingMatrix );
  poseObject.matrixWorld.multiply( standingMatrix );
  camera.matrixWorldInverse.multiply( standingMatrixInverse );
```

Comment (fernandojsg) [reply src/renderers/webvr/WebVRManager.js]: As the `standingMatrix` and the `standingMatrixInverse` won't be modified in the `!stageParams` case, you could just compute them once and just use them in that case, instead of getting the inverse and so on on each frame, but that's something that could be optimized later once we get it working ^_^

Comment (dmarcos) [reply src/renderers/webvr/WebVRManager.js]: scope.userHeight might change though

Comment (fernandojsg) [reply src/renderers/webvr/WebVRManager.js]: @dmarcos aps, so in that case skip my last comment :-P

**Comment by dmarcos** [comment]
I incorporate the changes suggested by @fernandojsg and did a complete round of testing on Vive, Rift and Windows MR

**Review by fernandojsg — APPROVED** [review]

**Comment by fernandojsg** [comment]
cool! lgtm

**Comment by dmarcos** [comment]
@mrdoob Mergea! Mergea! 

mergea

**Comment by arturitu** [comment]
I was testing this new reintroduction with 3DoF headsets and I think we are adding scope.userHeight twice. Because poseObject is camera and we are adding 'poseObject.position.set( 0, scope.userHeight, 0 );' when pose.position is null and besides, we are adding the standingMatrix that is 'standingMatrix.makeTranslation(0, scope.userHeight, 0);' when stageParameters are null.

**Comment by arturitu** [comment]
To try to clarify it, because now doesn't work as expected in all cases:

If we have the world position of a camera placed on for example: x=0, y=1, z=0, whether or not it has a parent. What is the expected behaviour in each case?

If we start from this proposal where we place the camera to render correctly located for Desktop/Magic Window/3DoF:
image
...we would need to add something outside the WebXRManager to have a correct camera position on 6DoF headset, if not we would have the camera at 1 + 1.6 meters (or the real height of the user if we have stageParameters) from the floor.

Because I think doesn't have sense to start with the camera at 0,0,0. Or am I missing something?

**Comment by dmarcos** [comment]
@arturitu I removed the 2nd branch of the if statement below that it is redundant and ends up doubling the height in 3DOF headsets with no stageParameters that seems to be the case in daydream. 

````javascript
if ( pose.position !== null ) {
     
     poseObject.position.fromArray( pose.position );

} else {

    poseObject.position.set( 0, scope.userHeight, 0 );

}
````
````

**Comment by dmarcos** [comment]
@arturitu In your schema above the height is always given by `userHeight` if a headset does not have `stageParameters`. That will cover all the use cases where a headset pose is available: 3DOF, 6DOF, Magic Window. For the 2D / Desktop case the `WebVRManager` plays no part and the height has to be set by the application. A-Frame covers that use case by saving the position on the camera when entering VR and restoring it when exiting. If you do something like this:
`<a-camera position="0 1.6 0">` The camera will be placed at 1.6m from the ground in all the cases you describe if there are no stageParameters. In that case the `sittingToStandingTransform` will be used. I hope it is a bit more clear.

File: src/renderers/webvr/WebVRManager.js [hunk 3]
```diff
@@ -121,6 +122,18 @@ function WebVRManager( renderer ) {
...
+		
+		if (stageParameters) {
+		  standingMatrix.fromArray( stageParameters.sittingToStandingTransform );
+		} else {
+		  standingMatrix.makeTranslation(0, scope.userHeight, 0);
+		}
+
+		standingMatrixInverse.getInverse( standingMatrix );
+		poseObject.matrixWorld.multiply( standingMatrix );
+		camera.matrixWorldInverse.multiply( standingMatrixInverse );
```
Comment (mrdoob) [inline src/renderers/webvr/WebVRManager.js]: Why `camera.matrixWorldInverse` instead of `poseObject.matrixWorldInverse` here?

Comment (dmarcos) [reply src/renderers/webvr/WebVRManager.js]: `matrixWorldInverse` is used to calculate the camera projection if I remember correctly and Object3Ds like `poseObject` doed not have a matrixWorldInverse. I never fully understood this part. I’m ready to be educated about it 🙂

File: src/renderers/webvr/WebVRManager.js [hunk 4]
```diff
@@ -107,10 +112,6 @@ function WebVRManager( renderer ) {
 
 			poseObject.position.fromArray( pose.position );
 
-		} else {
-
-			poseObject.position.set( 0, 0, 0 );
```
Comment (mrdoob) [inline src/renderers/webvr/WebVRManager.js]: Why did you remove this?

File: src/renderers/webvr/WebVRManager.js (hunk 3 above)
Comment (mrdoob) [reply src/renderers/webvr/WebVRManager.js]: > `matrixWorldInverse` is used to calculate the camera projection

Hmm... Not really. Did you added that line because if was needed in a specific situation? Otherwise if it's not needed I don't think we should add it. If that line is not needed, we can also remove all the `standingMatrixInverse` code.

File: examples/js/vr/ViveController.js [hunk 5]
```diff
@@ -71,6 +72,7 @@ THREE.ViveController = function ( id ) {
 			if ( pose.position !== null ) scope.position.fromArray( pose.position );
 			if ( pose.orientation !== null ) scope.quaternion.fromArray( pose.orientation );
 			scope.matrix.compose( scope.position, scope.quaternion, scope.scale );
+			scope.matrix.multiplyMatrices( scope.standingMatrix, scope.matrix );
```
Comment (mrdoob) [inline examples/js/vr/ViveController.js]:
```js
scope.matrix. premultiply( scope.standingMatrix );
```

File: src/renderers/webvr/WebVRManager.js (hunk 4 above)
Comment (dmarcos) [reply src/renderers/webvr/WebVRManager.js]: I think it is redundant since I've never seen a null position because it's either emulated via head/neck model or coming from the positional tracking. If null is a case we want to cover we should do it here as well:
https://github.com/mrdoob/three.js/pull/13158/files#diff-fdee169d721b24ed863f98ee8b9f67f4R117

File: src/renderers/webvr/WebVRManager.js (hunk 3 above)
Comment (dmarcos) [reply src/renderers/webvr/WebVRManager.js]: I did not add that code myself. It was there prior to the standingMatrix removal. I just ported the old logic. It works with it, I have not tried without. Happy to remove.
https://github.com/mrdoob/three.js/commit/a3f33be18c48124c3d5a0faf366685bde72587e6#diff-fdee169d721b24ed863f98ee8b9f67f4L146

File: src/renderers/webvr/WebVRManager.js (hunk 4 above)
Comment (mrdoob) [reply src/renderers/webvr/WebVRManager.js]: I understand. Can we leave that for another PR?

File: src/renderers/webvr/WebVRManager.js (hunk 3 above)
Comment (mrdoob) [reply src/renderers/webvr/WebVRManager.js]: Right, but now we are using `poseObject` and that old code doesn't make sense. Please remove the code if it's not needed. Lets try to have things under control.

File: examples/webvr_sandbox.html [hunk 6]
```diff
@@ -163,6 +172,9 @@
 				mesh.rotation.x = time * 2;
 				mesh.rotation.y = time * 5;
 
+				box.position.setFromMatrixPosition(poseTarget.matrixWorld);
```
Comment (mrdoob) [inline examples/webvr_sandbox.html]: What is this now? 😵

Comment (dmarcos) [reply examples/webvr_sandbox.html]: Oh that's @brianpeiris remanent code from his PR. removing

**Comment by mrdoob** [comment]
Thanks!

**Comment by dmarcos** [comment]
🚀 🚀 🚀

File: examples/webvr_sandbox.html (hunk 6 above)
Comment (brianpeiris) [reply examples/webvr_sandbox.html]: Oh, whoops. Thought I removed this myself. Thanks for cleaning up.

**Comment by danrossi** [comment]
The original PR code I tested against the VRController project and it was working with Daydream controller.  

However after undoing my changes and pulling the PR merge. now I can't see the controller object anymore unless I comment out the standing matrix any ideas ? 

https://jsfiddle.net/danrossi303/13onbkt0/
https://dev.electroteque.org/vrcontroller/raycast.html
Without

https://jsfiddle.net/danrossi303/gyzcacdm/1/
https://dev.electroteque.org/vrcontroller/raycast2.html

**Comment by danrossi** [comment]
With Oculus controller , it needs the standing matrix. But if I stand up the controller is above the head like with Gear VR / Daydream. Have to be sitting down for it to work. This was working with the original PR code. Is there extra requirements now ?

**Comment by danrossi** [comment]
With Oculus when I tilt up everything moves upwards including the box and controller. The controller is at the same position as DayDream view when that happens. When I tilt head down the controller is in the right position again.

**Comment by danrossi** [comment]
Yes if I head tilt up with the vive drag demo it does the same thing with Oculus.

**Comment by dmarcos** [comment]
@danrossi Thanks for giving it a try. Can you try this branch and let me know if it works for you? https://github.com/dmarcos/three.js/tree/r89fixGood
It has an older version of the PR without the following patch:
````
--- a/src/renderers/webvr/WebVRManager.js
+++ b/src/renderers/webvr/WebVRManager.js
@@ -112,7 +112,7 @@ function WebVRManager( renderer ) {

                        poseObject.position.fromArray( pose.position );

-               }
+               }

                if ( pose.orientation !== null ) {

@@ -120,10 +120,8 @@ function WebVRManager( renderer ) {

                }

-               poseObject.updateMatrixWorld();
-
                var stageParameters = device.stageParameters;
-
+
                if (stageParameters) {
                  standingMatrix.fromArray( stageParameters.sittingToStandingTransform );
                } else {
@@ -131,7 +129,8 @@ function WebVRManager( renderer ) {
                }

                standingMatrixInverse.getInverse( standingMatrix );
-               poseObject.matrixWorld.multiply( standingMatrix );
+               poseObject.position.applyMatrix4( standingMatrix );
+               poseObject.updateMatrixWorld();
                camera.matrixWorldInverse.multiply( standingMatrixInverse );

                if ( device.isPresenting === false ) return camera;
````

**Comment by danrossi** [comment]
Both versions with Oculus. If I launch webvr sitting down then stand up or launch while standing up. The floor and the controller is above my head. Maybe twice the height ? 

I'll have to double check gearVR

I am suffering random bugs with my headset maybe its mine only ?

**Comment by dmarcos** [comment]
@danrossi that's strange I cannot reproduce the problem you describe. Can I ask you a favor? Can you try supermedium and let me know if it works for you? https://www.supermedium.com/
It uses the same logic as https://github.com/dmarcos/three.js/tree/r89fixGood and it works fine on Oculus.

**Comment by danrossi** [comment]
It's doing the same thing in the browser although I am suffering wierd bugs with occulus not launching correctly and the openvr manager crashing. 

When I stand up the menu in that browser moves the same distance up and so does the controller.

**Comment by danrossi** [comment]
Both branches produce the same result although yours not so much. its intermittent. I am not getting the whole floor and scene to move up when standing up. But a head tilt will produce intermittent results where the controller and laser pointer and the entire floor moves above your head.

**Comment by danrossi** [comment]
@dmarcos  I've run more tests with your build against my vr controller fiddles. It seems less intermittent with the controllers moving up when tilting up compared to the dev branch but will still do it eventually.

**Comment by dmarcos** [comment]
@danrossi I'm sorry. I cannot reproduce any of the issues you describe. Can you try the following A-Frame example? https://aframe.io/aframe/examples/showcase/tracked-controls/

**Comment by danrossi** [comment]
@dmarcos Your branch fixes a height offset problem for GearVR / Daydream view. I've been running tests in that for now again. 

original
https://jsfiddle.net/danrossi303/13onbkt0/
https://dev.electroteque.org/vrcontroller/raycast.html

yours

https://jsfiddle.net/danrossi303/orfktz0h/1/
https://dev.electroteque.org/vrcontroller/raycast2.html

As for Oculus. 

There is definitely an intermittent problem with my Oculus but it might be related to the flickering display bugs I am experiencing and have had a hard time getting to the bottom of it. Might be a faulty device and I wish their support just get straight to the point.  It might be a fault on my end. 

Definitely just yesterday when I tilt up  the controller and floor will intermittently be above head height. Now it's not doing that.  just standing up is now doing that. If I tilt down the controller and floor move down. look ahead its above my head. 

My tests were with  the vive drag tests not just with the VRController project. 

http://localhost:8080/examples/?q=webvr#webvr_vive_dragging

It doesn't happen with GearVR / Daydream. 

I just tested the drag example in your branch. And the floor moves up when I stand up. The experience keeps changing and might be faults with my device but its consistent. 

The tracked controls demo . If i stand up from sitting down the object moves. sitting down the object is very close then when standing up its further away. When I stand up and tilt down the object is very close. The object is like "20cm" from the view. 

different to the controller and floor moving up.

**Comment by dmarcos** [comment]
@danrossi I think there’s something strange with your setup / device. Can you run native content without issue? Do you have access to a second Rift to compare with?

**Comment by danrossi** [comment]
I don't have a rift. I have those 3 devices only. 

Are you able to take a look at the differences in those examples with Daydream / Gear. Your branch fixes issues there. It's basically doing the same thing in those headsets I am experiencing with Oculus when standing up. 

Sorry about the noise. I may still have a faulty device and it's been terrible to deal with I have sent them logs multiple times. I have screen  flicker glitches when the headset is connected and presentation launching failures.  But there might be position data bugs which is causing that problem ? 

If I stay seated I can use the controller and laser pointer and raycast fine.

**Comment by danrossi** [comment]
@dmarcos Do I copy out your version of the webvrmanager ? I honestly don't want to spend too much time trying to figure out the occulus issue. But others might be able to confirm. 

I just need the fix for GearVR / Daydream for now. Unless I comment it out, using the standingmatrix with those headsets the floor is raised with the dev branch.

**Comment by danrossi** [comment]
after some tests this fixgood branch affects positioning of objects added to the camera that are repositioned on raycaster distance. Whereas the dev branch does not. 

I will have to now create work arounds with the controller to not use standingmatrix for daydream / gear if detected. Whereas this other branch is unaffected. I'm not across what the differences are.

**Comment by carstenschwede** [comment]
Can confirm what @danrossi and @arturitu are reporting in R90 and R91 for the Rift. Position in y is too high (probably twice the userheight).

R89 did not exhibit that behaviour.

**Comment by danrossi** [comment]
I gave up with that so I can keep going.  This went nowhere. Just don't stand up  when testing things. !
